server/store/cached_store.go

```diff
@@ -0,0 +1,420 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See License for license information.
+
+package store
+
+// TODO <>/<> wrap all errors
+
+import (
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	pluginapi "github.com/mattermost/mattermost-plugin-api"
+	"github.com/mattermost/mattermost-plugin-api/cluster"
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/plugin"
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-plugin-apps/server/incoming"
+	"github.com/mattermost/mattermost-plugin-apps/utils"
+)
+
+const (
+	CachedStoreEventID      = "cached_store"
+	CachedStorePutMethod    = "put"
+	CachedStoreDeleteMethod = "delete"
+)
+
+type cachedStoreEvent[T any] struct {
+	Key       string    `json:"key"`
+	Method    string    `json:"method"`
+	SentAt    time.Time `json:"sent_at"`
+	StoreName string    `json:"name"`
+
+	Data T `json:"data,omitempty"`
+}
+
+type IndexEntry[T any] struct {
+	Key       string `json:"k"`
+	ValueHash string `json:"h"`
+	data      T
+}
+
+type StoredIndex[T any] struct {
+	Data []IndexEntry[T]
+}
+
+type Index[T any] map[string]*IndexEntry[T]
```

Comment:
Well, there's no real external API since these are wrapped in `AppStore` and the likes, but I am making some types public with the intent of go-documenting them.